Always audit GitHub license

dtrodrigues opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
The current logic only checks the GitHub license for new formulae. This change will always check for the license GitHub for formulae hosted on GitHub in case the license has changed and none of the Homebrew licenses for a formula match the license in GitHub. This logic detected the issues in https://github.com/Homebrew/homebrew-core/pull/59728.

This also excludes checking versioned formulae because the GitHub API only returns the license for the default branch and the license information for a versioned branch may be different (see `scw@1` vs `scw`).

Rylan12 commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -380,9 +380,11 @@ def audit_license
+        user, repo = get_repo_data(%r{https?://github\.com/([^/]+)/([^/]+)/?.*})
         return if user.blank?
 
+        return if formula.versioned_formula?
```
Maybe move this up to be with the `return unless @online`? No need to fetch repo data if we already know we won't use it.

MikeMcQuaid commented:
I think we should check existing licenses on existing formulae with licensees and require licenses on new formulae but not check/require/fail on existing formulae without licenses. We should not check/complain about "deprecated" license use on non-`--strict`.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -378,9 +378,9 @@ def audit_license
         end
 
-        return unless @online
+        return unless @online && !formula.versioned_formula?
```
Why `!formula.versioned_formula?`? Also, please refactor to not use `unless .. &&` (or `unless .. ||`)

Rylan12 replied:
> This also excludes checking versioned formulae because the GitHub API only returns the license for the default branch and the license information for a versioned branch may be different (see `scw@1` vs `scw`).

MikeMcQuaid replied:
Cool. In that case: I think this warrants a comment.

Rylan12 commented:
Another issue came up in https://github.com/Homebrew/homebrew-core/pull/59728#issuecomment-674609197. It's possible that the GitHub license listed may not match what we actually ship (like is the case for `cmockery`). For these instances, it might be good to have an allowlist for formulae that we know have different licenses than what the GitHub API says.

Bo98 commented:
Yeah the GitHub API only supports HEAD of the default branch unfortunately - no tags.

In the ideal world, we'd download the archive and check the license file (e.g. [via this](https://github.com/licensee/licensee)), but this is trickier to do in an audit.

If we consistently installed metafiles, a Cellar check would have been possible but I don't think we do.

MikeMcQuaid commented:
> For these instances, it might be good to have an allowlist for formulae that we know have different licenses than what the GitHub API says.

Agreed.



> In the ideal world, we'd download the archive and check the license file (e.g. [via this](https://github.com/licensee/licensee)), but this is trickier to do in an audit.

That's what the batch import script did, out of interest.

MikeMcQuaid commented:
> As long as we don't forget to remove it when upstream cuts a release with the license change and that people know to always double check when they see a license mismatch audit.

In that case: make it name and version.

dtrodrigues commented:
Based on the feedback, I've updated to have a version-specific formula allowlist for mismatches. Since this would also allow versioned formulae to be accounted for, I removed the check for if formulae are versioned.
